ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,151 @@
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE_MILLISECONDS", 0))
+
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        if apm_flush_deadline < 200:
+            log.warning(
+                "DD_APM_FLUSH_DEADLINE_MILLISECONDS was set to %d, this value will be overridden to 200ms instead.",
+                apm_flush_deadline,
+            )
+            apm_flush_deadline = 200
+
+        remaining_time_in_millis = apm_flush_deadline
+
+    # Subtracting 100 milliseconds to ensure we have time to flush.
+    # TODO: Update logic to calculate an approximate of hong long it will
+    # take us to flush the spans on the queue.
+    remaining_time_in_seconds = max((remaining_time_in_millis - 100) / 1000, 0)
+    signal.setitimer(signal.ITIMER_REAL, remaining_time_in_seconds)
+
+
+def datadog_instrumentation(func, args, kwargs):
+    """Patches an AWS Lambda handler function for Datadog instrumentation."""
+    context = get_argument_value(args, kwargs, -1, "context")  # context is always the last parameter
+    check_timeout(context)
+
+    return func(*args, **kwargs)
+
+
+def modify_module_name(module_name):
+    """Returns a valid modified module to get imported."""
+    return ".".join(module_name.split("/"))
+
+
+def get_handler_and_module():
+    """Returns the user AWS Lambda handler and module."""
+    path = os.environ.get("DD_LAMBDA_HANDLER", None)
+    if path is None:
+        from datadog_lambda.wrapper import datadog_lambda_wrapper
+
+        handler = getattr(datadog_lambda_wrapper, "__call__")
+
+        def wrapper(func, args, kwargs):
+            return datadog_instrumentation(func, args, kwargs)
+
+        return handler, datadog_lambda_wrapper, wrapper
+    else:
+        parts = path.rsplit(".", 1)
+        (mod_name, handler_name) = parts
+        modified_mod_name = modify_module_name(mod_name)
+        handler_module = import_module(modified_mod_name)
+        handler = getattr(handler_module, handler_name)
+        return handler, handler_module, datadog_instrumentation
+
+
+def _has_patch_module():
+    """
+    Ensures that the `datadog_lambda` integration can be patched.
+
+    It checks either the user has the DD_LAMBDA_HANDLER set correctly.
+    Or if the `datadog_lambda` package is installed.
+    """
+    path = os.environ.get("DD_LAMBDA_HANDLER", None)
+    if path is None:
+        try:
+            import_module("datadog_lambda.wrapper")
+        except Exception:
+            return False
+    else:
+        parts = path.rsplit(".", 1)
+        if len(parts) != 2:
+            return False
+    return True
+
+
+def _in_aws_lambda():
```

Comment:
I think we have this same check else where in the library: https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/tracer.py#L1067. It might be worth it to consolidate them in a new `ddtrace.internal.serverless` module. WDYT?

Reply:
Sounds really great, for general stuff it will be used, since we won't have other clouds yet, I'll just add them in the `__init__.py` file